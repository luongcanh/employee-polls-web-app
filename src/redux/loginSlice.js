import { createSlice } from '@reduxjs/toolkit';

const initialState = {};

export const loginSlice = createSlice({
	name: 'login',
	initialState,
	reducers: {
		setLogin: (state, action) => {
			state.authedUser = action.payload.id;
		},
		setLogout: (state) => {
			state.authedUser = null;
		}
	},
});

export const { setLogin, setLogout } = loginSlice.actions;

export const selectLogin = (state) => state?.login?.authedUser;

export default loginSlice.reducer;
