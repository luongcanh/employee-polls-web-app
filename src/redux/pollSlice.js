import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';
import { getAllPolls } from './utils/callAPI';
import { ANSWER_OPTION } from '../constants/constant';
import { checkIfAnswered } from './utils/common';
import { _saveQuestion, _saveQuestionAnswer } from '../_DATA';

const initialState = {
	loading: false,
	polls: [],
};

/**
 * @description Get all polls
 */
export const fetchPolls = createAsyncThunk('polls/getAllPolls', async () => await getAllPolls());

export const handleSubmitAnswer = createAsyncThunk(
	'polls/handleSubmitAnswer',
	async ({ authedUser, qid, answer }) => {
		await _saveQuestionAnswer({ authedUser, qid, answer });

		return { authedUser, qid, answer };
	},
);

/**
 * @description Create new poll
 */
export const handleCreateQuestion = createAsyncThunk(
	'polls/handleCreateQuestion',
	async ({ author, optionOneText, optionTwoText }) => {
		const result = await _saveQuestion({ author, optionOneText, optionTwoText });

		return result;
	},
);

const pollSlice = createSlice({
	name: 'polls',
	initialState,
	reducers: {},
	extraReducers: (builder) => {
		builder.addCase(fetchPolls.pending, (state) => {
			state.loading = true;
		});
		builder.addCase(fetchPolls.fulfilled, (state, action) => {
			state.polls = action.payload;
			state.loading = false;
		});
		builder.addCase(handleSubmitAnswer.pending, (state) => {
			state.loading = true;
		});
		builder.addCase(handleSubmitAnswer.fulfilled, (state, action) => {
			let currentVoters = state.polls[action.payload.qid][action.payload.answer].votes;
			if (
				!checkIfAnswered(state.polls[action.payload.qid], action.payload.authedUser) &&
				(action.payload.answer === ANSWER_OPTION.OPTION_ONE ||
					action.payload.answer === ANSWER_OPTION.OPTION_TWO)
			) {
				currentVoters.push(action.payload.authedUser);
			}
			state.loading = false;
		});
		builder.addCase(handleCreateQuestion.pending, (state) => {
			state.loading = true;
		});
		builder.addCase(handleCreateQuestion.fulfilled, (state, action) => {
			state.loading = false;
			state.polls[action.payload.id] = action.payload;
		});
		builder.addCase(handleCreateQuestion.rejected, (state, action) => {
			state.loading = false;
		});
	},
});

export default pollSlice.reducer;
