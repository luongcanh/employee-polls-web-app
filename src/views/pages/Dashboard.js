import React, { useEffect, useState } from 'react';
import { Box, Tab, Tabs, Container } from '@mui/material';
import QuestionsList from '../components/QuestionsList';
import { useSelector } from 'react-redux';
import { ANSWER_TYPE } from '../../constants/constant';

const Dashboard = () => {
	const [value, setValue] = useState(0);
	const authedUser = useSelector((state) => state.login.authedUser);
	const [newQuestions, setNewQuestions] = useState({});
	const [answeredQuestions, setAnsweredQuestions] = useState({});
	const allQuestions = useSelector((state) => state.polls.polls);
	const users = useSelector((state) => state.users.users);

	useEffect(() => {
		if (authedUser) {

			const answeredQuestionIds = Object.keys(users[authedUser].answers).sort(
				(firstQId, secondQId) => {
					return allQuestions[secondQId].timestamp - allQuestions[firstQId].timestamp;
				},
			);

			const newQuestionIds = Object.keys(allQuestions)
				.filter((questionId) => !users[authedUser].answers[questionId])
				.sort((firstQId, secondQId) => {
					return allQuestions[secondQId].timestamp - allQuestions[firstQId].timestamp;
				});
			
			setNewQuestions(
				newQuestionIds.reduce((acc, questionId) => {
					return { ...acc, [questionId]: allQuestions[questionId] };
				}, {}),
			);
			
			setAnsweredQuestions(
				answeredQuestionIds.reduce((acc, questionId) => {
					return { ...acc, [questionId]: allQuestions[questionId] };
				}, {}),
			);
		}
		// eslint-disable-next-line
	}, []);

	/**
	 * @description Handle tab selection changes (New Question/ Done)
	 */
	const handleOnChange = (event, newValue) => {
		setValue(newValue);
	};

	return (
		<div>
			<Container maxWidth="xl">
				<Box sx={{ width: '100%' }}>
					<Box sx={{ borderBottom: 1, borderColor: 'divider' }}>
						<Tabs value={value} onChange={handleOnChange}>
							<Tab label="New Question" />
							<Tab label="Done" />
						</Tabs>
					</Box>
					{value === 0 && (
                        <QuestionsList questions={newQuestions} questionType={ANSWER_TYPE.UNANSWERED} />
                    )}
                    {value === 1 && (
                        <QuestionsList questions={answeredQuestions} questionType={ANSWER_TYPE.ANSWERED} />
                    )}
				</Box>
			</Container>
		</div>
	);
};

export default Dashboard;
