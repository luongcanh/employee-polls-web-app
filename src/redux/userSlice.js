import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';
import { getAllUsers } from './utils/callAPI';
import { handleCreateQuestion, handleSubmitAnswer } from './pollSlice';

const initialState = {
    loading: false,
    users: [],
};

export const fetchUsers = createAsyncThunk('users/getAllUsers', async () => await getAllUsers());

const userSlice = createSlice({
    name: 'users',
    initialState,
    reducers: {},
    extraReducers: (builder) => {
        builder.addCase(fetchUsers.pending, (state) => {
            state.loading = true;
        });
        builder.addCase(fetchUsers.fulfilled, (state, action) => {
            state.users = action.payload;
            state.loading = false;
        });
        builder.addCase(handleSubmitAnswer.pending, (state) => {
            state.loading = true;
        });
        builder.addCase(handleSubmitAnswer.fulfilled, (state, action) => {
            state.users[action.payload.authedUser].answers[action.payload.qid] = action.payload.answer;
            state.loading = false;
        });
        builder.addCase(handleCreateQuestion.pending, (state) => {
            state.loading = true;
        });
        builder.addCase(handleCreateQuestion.fulfilled, (state, action) => {
            state.loading = false;
            state.users[action.payload.author].questions.push(action.payload.id);
        });
        builder.addCase(handleCreateQuestion.rejected, (state, action) => {
            state.loading = false;
        });
    },
});

export default userSlice.reducer;
