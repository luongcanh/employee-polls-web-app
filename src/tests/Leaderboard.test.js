/* eslint-disable testing-library/prefer-screen-queries */
import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';
import { Provider } from 'react-redux';
import Dashboard from '../views/pages/Dashboard';
import { store } from '../redux/store';
import { setLogin } from '../redux/loginSlice';
import { fetchUsers } from '../redux/userSlice';
import { BrowserRouter as Router } from 'react-router-dom';
import { Leaderboard } from '@mui/icons-material';

describe('Leaderboard Component', () => {
	let component;

	beforeEach(async () => {
		await store.dispatch(fetchUsers());
		await store.dispatch(setLogin({ id: 'sarahedo' }));
		// eslint-disable-next-line
		component = render(
			<Provider store={store}>
				<Router>
					<Leaderboard />
				</Router>
			</Provider>,
		);
	});

	it('renders Leaderboard component correctly', () => {
        expect(component).toBeDefined();
        expect(component).toMatchSnapshot();
	});

});
