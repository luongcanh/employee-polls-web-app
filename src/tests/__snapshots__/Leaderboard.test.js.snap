// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Leaderboard Component renders Leaderboard component correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="LeaderboardIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M7.5 21H2V9h5.5zm7.25-18h-5.5v18h5.5zM22 11h-5.5v10H22z"
        />
      </svg>
    </div>
  </body>,
  "container": <div>
    <svg
      aria-hidden="true"
      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
      data-testid="LeaderboardIcon"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M7.5 21H2V9h5.5zm7.25-18h-5.5v18h5.5zM22 11h-5.5v10H22z"
      />
    </svg>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
